phastSim/phastSim.py

```diff
@@ -641,23 +809,196 @@ def sampleMutation(self, allele, rates, rand):
+            node.deletionRate = self.deletionRate[-1]
+            node.deletionLength = self.deletionLength[-1]
+            node.insertionLength = self.insertionLength[-1]
+
+            node.rate = -1 * node.rates[node.allele][node.allele]
+            
+
+
+            return node
+
+        else:
+            # cut the list in half and build a tree for each half
+            middle = pos_left + int((pos_right - pos_left) / 2)
+            half_length = int((pos_right - pos_left)/2)
+
+            child0 = self.buildTree(genomeNode(level=node.level), insert_list[:half_length+1], pos_left, middle)
+            child1 = self.buildTree(genomeNode(level=node.level), insert_list[half_length+1:], middle+1, pos_right)
+            node.belowNodes = [child0, child1]
+            node.rate = child0.rate + child1.rate
+            node.genomePos = [(self.n_insertions, pos_left), (self.n_insertions, pos_right)]
+            return node
+
+
+    def sampleInsertion(self, rate, node, rand, level):
+
+        # sample a geometric random variable
+        insertion_length = 1 + int(np.log(rand)/np.log(rate))
+        
+        # TODO generate a nucleotide sequence using the correct model
+        insertion = np.random.choice(self.allelesList, insertion_length, self.insertionFrequencies)
+
+        print("*************", insertion)
+        # turn the nucleotide sequence into a genomeTree
+        genomeSeq = "".join(insertion)
+
+        newNode = genomeNode(level=level)
+
+        # generate the genome tree of this insert from (node, genomeSeq)
+        subTree = self.buildTree(newNode, genomeSeq, 0, len(genomeSeq) - 1)
+
+        self.normalizeRates(subTree, self.indels, self.codon, len(genomeSeq), self.scale)
+
+        # build a bigger tree of the form t = (node, subTree). 
+        newRoot = genomeNode(level=level)
+        newRoot.belowNodes = [node, subTree]
+        newRoot.rate = node.rate + subTree.rate
+
+
+        mutData = mutation(mType=mType.INS, genomePos=node.refNode.genomePos[0], insert=genomeSeq, index=self.n_insertions)
+
+        return mutData, newRoot
 
     def findPos(self, rand, parentGenomeNode, level):
         # find position to mutate along the genome, and update temporary genome tree structure as you go
         if parentGenomeNode.isTerminal:
             # reached a terminal node, now sample the mutation event at the position and update all rates
             node = parentGenomeNode.refNode
             a = parentGenomeNode.allele
+            
+            # we'll choose an indel if rand is less than a threshold determined by the insertion and deletion rates for this site
+            if self.indels:
+
+                # deletions
+                if rand < node.deletionRate:
+
+                    attempted_deletion_length = self.sampleDeletion(node.deletionLength, rand/node.deletionRate)
+                    pos = node.genomePos[0]
+                    parentGenomeNode.rate = node.rate
```

Comment:
What is the purpose of this line? It seems to me that it should be removed, as it reverts the rate of the current node to the one at the start of the simulations - it doesn't seem right.

Reply:
I know what I was trying to do here, but maybe have not done it correctly. 

Because findPos is recursive, and the parent genomeNodes of this node get their rates updated based on their children, the terminal node needed to have a rate property, so I had to set it as something. Originally I tried setting it to 0 but I thought that wouldn't work, because when it came to tidying up the genome tree by calling deleteNodes (with the same random number), the new tree that it would get called on would no longer have the same rates as before, they would be updated to different numbers, and so you could no longer search for a position in the genome tree using the same rand number. What I was trying to do was to keep the rate for this new node unchanged compared to how it was before, and then update the tree (deleting the nodes that needed to be deleted) with deleteNodes. Does this make sense to you? I did not intend to update the rate of the node to how it was at the start of the simulation, I just wanted it to remain unchanged compared to the original node, basically how it was before findPos was called...

Reply:
to fix this, it should be ok to just replace this line with a new line at position 1083 that says:

newChild.rate=child.rate